runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java

```diff
@@ -0,0 +1,210 @@
+      throw new RuntimeException(e);
+    }
+    KvCoder<K, V> inputKvCoder = (KvCoder<K, V>) inputCoder.getValueCoder();
+    Coder<K> inputKeyCoder = inputKvCoder.getKeyCoder();
+    Coder<V> inputValueCoder = inputKvCoder.getValueCoder();
+    WindowingStrategy windowingStrategy = getWindowingStrategy(inputId, components);
+    WindowFn<Object, BoundedWindow> windowFn = windowingStrategy.getWindowFn();
+    WindowedValue.WindowedValueCoder<V> wvCoder =
+        WindowedValue.FullWindowedValueCoder.of(inputValueCoder, windowFn.windowCoder());
+
+    JavaRDD<WindowedValue<KV<K, Iterable<V>>>> groupedByKeyAndWindow;
+    if (windowingStrategy.getWindowFn().isNonMerging()
+        && windowingStrategy.getTimestampCombiner() == TimestampCombiner.END_OF_WINDOW) {
+      // we can have a memory sensitive translation for non-merging windows
+      groupedByKeyAndWindow =
+          GroupNonMergingWindowsFunctions.groupByKeyAndWindow(
+              inputRdd, inputKeyCoder, inputValueCoder, windowingStrategy);
+    } else {
+      Partitioner partitioner = getPartitioner(context);
+      JavaRDD<WindowedValue<KV<K, Iterable<WindowedValue<V>>>>> groupedByKeyOnly =
+          GroupCombineFunctions.groupByKeyOnly(inputRdd, inputKeyCoder, wvCoder, partitioner);
+      // for batch, GroupAlsoByWindow uses an in-memory StateInternals.
+      groupedByKeyAndWindow =
+          groupedByKeyOnly.flatMap(
+              new SparkGroupAlsoByWindowViaOutputBufferFn<>(
+                  windowingStrategy,
+                  new TranslationUtils.InMemoryStateInternalsFactory<>(),
+                  SystemReduceFn.buffering(inputValueCoder),
+                  context.serializablePipelineOptions,
+                  AggregatorsAccumulator.getInstance()));
+    }
+    context.pushDataset(getOutputId(transformNode), new BoundedDataset<>(groupedByKeyAndWindow));
+  }
+
+  private static <InputT, OutputT> void translateExecutableStage(
+      PTransformNode transformNode, RunnerApi.Pipeline pipeline, SparkTranslationContext context) {
+
+    RunnerApi.ExecutableStagePayload stagePayload;
+    try {
+      stagePayload =
+          RunnerApi.ExecutableStagePayload.parseFrom(
+              transformNode.getTransform().getSpec().getPayload());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    String inputPCollectionId = stagePayload.getInput();
+    Dataset inputDataset = context.popDataset(inputPCollectionId);
+    JavaRDD<WindowedValue<InputT>> inputRdd = ((BoundedDataset<InputT>) inputDataset).getRDD();
+    Map<String, String> outputs = transformNode.getTransform().getOutputsMap();
+    BiMap<String, Integer> outputMap = createOutputMap(outputs.values());
+
+    SparkExecutableStageFunction<InputT> function =
+        new SparkExecutableStageFunction<>(stagePayload, context.jobInfo, outputMap);
+    JavaRDD<RawUnionValue> staged = inputRdd.mapPartitions(function);
+
+    for (String outputId : outputs.values()) {
```

Comment:
I found there is an issue here--sometimes (e.g. for PAsserts) there will be no outputs, in which case we need to create a sink (like [this](https://github.com/ibzib/beam/blob/1f1660f0c97e65bb91e2c5aa7694b6d7b9215bbd/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java#L398-L403)). I will add a fix in a future PR.